unified-runtime/source/adapters/offload/queue.hpp

```diff
@@ -15,7 +15,35 @@
 
 #include "common.hpp"
 
+constexpr size_t OOO_QUEUE_POOL_SIZE = 32;
+
 struct ur_queue_handle_t_ : RefCounted {
-  ol_queue_handle_t OffloadQueue;
+  ur_queue_handle_t_(ol_device_handle_t Device, ur_queue_flags_t Flags)
+      : OffloadQueues((Flags & UR_QUEUE_FLAG_OUT_OF_ORDER_EXEC_MODE_ENABLE)
+                          ? 1
+                          : OOO_QUEUE_POOL_SIZE),
+        QueueOffset(0), OffloadDevice(Device), Flags(Flags) {}
+
+  // In-order queues only have one element here, while out of order queues have
+  // a bank of queues to use. We rotate through them round robin instead of
+  // constantly creating new ones in case there is a long-running program that
+  // never destroys the ur queue. Out-of-order queues create ol queues when
+  // needed; any queues that are not yet created are nullptr.
+  std::vector<ol_queue_handle_t> OffloadQueues;
+  size_t QueueOffset;
   ol_device_handle_t OffloadDevice;
+  ur_queue_flags_t Flags;
+
+  ol_result_t nextQueue(ol_queue_handle_t &Handle) {
```

Comment:
If we're going to use round-robin in-order queues, why not use the common `stream_queue_t` implementation like the CUDA and HIP adapters? This is much cleaner, but I worry that by the time we implement other queue related things, especially event wait lists, we'll end up needing to reimplement everything it does anyway.